### PR TITLE
Fix vectorized calculations on POWER

### LINKED ISSUE #57537: Nan propagation of min/max function broken for POWER
## 🐛 Bug

Running the tests shows a failure with nan propagation for `torch.max`/`torch.min` introduced in 1.8.0 via https://github.com/pytorch/pytorch/pull/41541

Example:

```
FAIL: test_min_max_binary_op_nan_cpu_float32 (__main__.TestBinaryUfuncsCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/easybuild-tmp/eb-TDt0D1/tmpeqDImk/lib/python3.8/site-packages/torch/testing/_internal/common_device_type.py", line 290, in instantiated_test
    result = test_fn(self, *args)
  File "test_binary_ufuncs.py", line 1306, in test_min_max_binary_op_nan
    self.assertTrue(torch.isnan(ma[i]), "max(a, b): {}, a: {}, b: {}".format(ma[i], a[i], b[i]))
AssertionError: tensor(False) is not true : max(a, b): 0.29149895906448364, a: nan, b: 0.29149895906448364
```

## To Reproduce

Run tests on PPC

See e.g https://gist.github.com/Flamefire/3f11af7f908742908d0d035d3e430646

## Expected behavior

Tests succeed

## Environment

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Any other relevant information: ppc9le

## Additional context

This failure was (knowingly) introduced by the mentioned PR.
Question now is: Is the behavior on NaNs defined by PyTorch? If so there is a bug which needs fixing. Otherwise the relevant tests should be removed or changed.

Also see e.g. https://eigen.tuxfamily.org/bz/show_bug.cgi?id=1494 / https://gitlab.com/libeigen/eigen/-/issues/1494

> using the vec_max intrinsic which follows IEEE-754R and thus returns numeric for numeric vs NaN comparisons.

If the NaN-propagating behavior is wanted, I don't see any other approach but not using `vec_min` etc on VSX

### LINKED ISSUE #59248: Multiple failures in test_unary_ufuncs on POWER
## 🐛 Bug

A large bunch of tests in test_unary_ufuncs fails on POWER9 after the vector intrinsic introduction in PyTorch 1.8 they look related hence pasted as a single issue.

## To Reproduce

1. run `test_unary_ufuncs.py` on PPC

## Expected behavior

Tests pass

## Environment

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): `CMAKE_BUILD_TYPE=Release BUILD_TEST=0 PYTORCH_BUILD_VERSION=1.8.1 PYTORCH_BUILD_NUMBER=1 MAX_JOBS=$(nproc) BLAS=Eigen USE_FFMPEG=1 BUILD_CUSTOM_PROTOBUF=0 USE_IBVERBS=1 USE_CUDA=0 USE_METAL=0 python setup.py install`
 - Python version: 3.8.6

## Additional context

I'll try to categorize the tests into failure types and (possible) explanations:

- test_contig_vs_every_other_angle_cpu_complex64: 256 / 513 elements failed `The greatest difference was 3.0628208592534065 (-3.096264600753784 vs. -0.033443741500377655)`: Looks like the VSX version gets bogus values here
-  test_contig_vs_every_other_angle_cpu_float32/float64, test_non_contig_angle_cpu_float32/float64, test_non_contig_expand_angle_cpu_int16/int32/int64/int8: `greatest difference was 3.1415927410125732 (0.0 vs. 3.1415927410125732)`: This is an omission: The VSX implementation does not return PI when the argument is negative: https://github.com/pytorch/pytorch/blob/afdfd2288ab839ece89e667e44777201536b2d82/aten/src/ATen/cpu/vec/vec256/vsx/vec256_double_vsx.h#L257
- test_non_contig_angle_cpu_complex128/complex64: `greatest difference was 1.5819721883714497 (0.8557325138535072 vs. 2.437704702224957)` and `greatest difference was 2.4029039442539215 (-0.4417212903499603 vs. -2.844625234603882)`: Similar to the first although the relation to (almost) pi/2 and 3/4*pi suggest some mixing of values
- test_non_contig_index_angle_cpu_complex128: `greatest difference was 0.38246162968485264 (-1.3715882372626278 vs. -1.7540498669474804)`
- test_reference_numerics_angle_cpu_complex128/complex64: Similar failures, probably the whole complex-angle stuff is bugged
- test_reference_numerics_angle_cpu_float32/float64: `greatest difference was nan (0.0 vs. nan)`, not sure here, maybe same as point 2
- test_reference_numerics_angle_cpu_int16/int32/int64/int8: Same as 2 (PI vs 0)
- test_reference_numerics_log1p_cpu_float32 (and log10, log2, log): Compares a value (88.72283935546875, 128.0, 38.53184127807617) against inf
- test_reference_numerics_logit_cpu_float32 `greatest difference was nan (-87.3365478515625 vs. nan)`
- test_sgn_cpu_complex128/complex64: `greatest difference was 0.5775589227244207 (0.8735823660790683 vs. 0.29602344335464753)` and `greatest difference was 0.38670969009399414 (-2.521059989929199 vs. -2.9077696800231934)`, it is the angle comparison that fails, so same as 1

cc @mruberry @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by Flamefire**
This fixes multiple bugs introduced by the VSX optimized code in https://github.com/pytorch/pytorch/pull/41541

- min/max/clamp now consistently return nan when any value is NaN as on other architectures
- The non-complex angle functions return PI for negative values now
- The complex angle functions have been corrected and optimized
- The float32-log function implementation returned a wrong result when inf was passed (and maybe other inputs), replaced by the sleef function just as for float64

Fixes #59248
Fixes #57537

### CODE REVIEW

**Comment by ezyang** [comment]
Since we don't have any POWER CI, do you mind explaining briefly how you tested the changes here? (Just for posterity)

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_int64_vsx.h [hunk 1]
```diff
@@ -1,8 +1,8 @@
 #pragma once
 
-#include <ATen/cpu/vec/vec256/intrinsics.h>
-#include <ATen/cpu/vec/vec256/vec256_base.h>
-#include <ATen/cpu/vec/vec256/vsx/vsx_helpers.h>
+#include <ATen/cpu/vec/Vectorized/intrinsics.h>
+#include <ATen/cpu/vec/Vectorized/Vectorized_base.h>
+#include <ATen/cpu/vec/Vectorized/vsx/vsx_helpers.h>
```
Comment (ezyang) [inline aten/src/ATen/cpu/vec/vec256/vsx/vec256_int64_vsx.h:5]: Huh, what's going on here?

Comment (Flamefire) [reply aten/src/ATen/cpu/vec/vec256/vsx/vec256_int64_vsx.h:5]: Ouch sorry. I made that patch with 1.8.1 and ported it to master where the vec256 class was renamed but not the paths (which I find confusing) will fix this.

**Comment by Flamefire** [comment]
> Since we don't have any POWER CI, do you mind explaining briefly how you tested the changes here? (Just for posterity)

I compiled as usual and then run `test_unary_ufuncs.py` and `test_binary_ufuncs.py` successfully which failed before (see the linked issues)

BTW: I found some missing test coverage: The NaN propagation of min/max and the clamp/clamp_min/clamp_max function is not fully covered. I.e. that min/max return NaN when either is NaN but clamp only returns NaN when the second is. So clamp_min!=max which I missed at first and it made another test fail with something hard to find. See https://github.com/pytorch/pytorch/pull/59382/commits/dd62c8970a41325c9ce88e759e2572db59d960d9

**Comment by Flamefire** [comment]
@ezyang I rebased onto master and recompiled that, then ran the 2 test files. All succeeds which hasn't before but getting 2 new failures on master:
- test_reference_numerics_hard_atanh_cpu_complex64
- test_complex_edge_values_cpu_complex64

Those refer to acos and atanh which dispatch to the std variants and those return nan/inf for the tested input values (on PPC and x86) as can be checked with:
```
#include <iostream>
#include <complex>

int main(){
  auto bar = std::acos(std::complex<float>(0, 1e+20));
  auto baz = std::atanh(std::complex<float>(-501, 1e+20));
  std::cout << bar << std::endl;
  std::cout << baz << std::endl;
}
```
--> `(1.5708,-inf) (-nan,1.5708)`

I'm not sure if this is changed on x86 already to dispatch to something else which would need to be done too on PPC, maybe @malfet can comment here? Seen him on the related issue https://github.com/pytorch/pytorch/issues/42952

**Comment by ezyang** [comment]
Seems OK to leave those fixes for later. acos seems related to https://github.com/pytorch/pytorch/issues/52310

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**